docs/design.md

```diff
@@ -0,0 +1,431 @@
+* `substitutions` - list, optional - A sequence of substitutions to apply. See
+  the Substitutions section for additional details.
+  * `dest` - dict, required - A description of the inserted content destination.
+    * `path` - string, required - The JSON path where the data will be placed
+      into the `data` section of this document.
+  * `src` - dict, required - A description of the inserted content source.
+    * `schema` - string, required - The `schema` of the source document.
+    * `name` - string, required - The `metadata.name` of the source document.
+    * `path` - string, required - The JSON path from which to extract data in
+      the source document relative to its `data` section.
+
+Here is a fictitious example of a complete document which illustrates all the
+valid fields in the `metadata` section.
+
+```yaml
+---
+schema: some-service/ResourceType/v1
+metadata:
+  schema: metadata/Document/v1
+  name: unique-name-given-schema
+  storagePolicy: cleartext
+  labels:
+    genesis: enabled
+    master: enabled
+  layeringDefinition:
+    abstract: true
+    layer: region
+    parentSelector:
+      labels: parents
+      must: have
+    actions:
+      - method: merge
+        path: .path.to.merge.into.parent
+      - method: delete
+        path: .path.to.delete
+  substitutions:
+    - dest:
+        path: .substitution.target
+      src:
+        schema: another-service/SourceType/v1
+        name: name-of-source-document
+        path: .source.path
+data:
+  path:
+    to:
+      merge:
+        into:
+          parent:
+            foo: bar
+          ignored:  # Will not be part of the resultant document after layering.
+            data: here
+  substitution:
+    target: null  # Paths do not need to exist to be specified as substitutiond destinations.
+```
+
+
+##### schema: metadata/Control/v1
+
+This schema is the same as the `Document` schema, except it omits the
+`storagePolicy`, `layeringDefinition`, and `substitutions` keys, as these
+actions are not supported on `Control` documents.
+
+The complete list of valid `Control` document kinds is specified below along
+with descriptions of each document kind.
+
+##### schema: metadata/Tombstone/v1
+
+The only valid key in a `Tombstone` metadata section is `name`.  Additionally,
+the top-level `data` section should be omitted.
+
+### Layering
+
+When queried, Deckhand assembles the chain of partial documents in the order
+specified in the `deckhand/LayeringPolicy/v1` document. Each chain of partial
+documents is made up only of documents using a single `schema`.
+
+<!-- TODO: describe document chain construction w/ diagrams -->
+
+During each layering step, the list of `actions` will be applied in order.
+Supported actions are:
+
+* `merge` - a "deep" merge that layers new and modified data onto existing data
+* `replace` - overwrite data at the specified path and replace it with the data
+  given in this document
+* `delete` - remove the data at the specified path
+
```

Comment:
Just restating all the questions and adding some related ones here for my own clarity:

1.  Is it an error to have additional data in a merge document that is not specified in at least one action?
2. Should we have a default behavior when no actions are specified?
3. If we have a default action, should we apply it even when there are other actions?
4. Do we allow action paths to overlap?
5. Do we allow multiple merge target parents from the same layer?

My thoughts on each:

1. I was initially thinking to ignore it, but maybe it should be an error?  I'm not 100% clear if this is a tricky one to check or not.
2. I don't think it's needed, assuming we allow paths to overlap (question 4), since in that case you can just add a `method=merge, path=.` action.
3. I'm torn on this, again relates to action path overlap.
4. I think there's not much harm in allowing this since it's fully deterministic and still has to pass schema validation after it occurs.
5. I would like to avoid allowing multiple targets from the same layer, since the main push for that seemed to be related to Armada using a single large document, and that's going to change soon.

I'm feeling a bit less confident about the merge chain construction at this point.  I would really like to avoid a proper DAG or even a tree and just have a clear, linear path (maybe specifying just parent name or similar).  I guess we could implement that by limiting the parent selection to the highest precedence (latest order) document that matches the selector (that is also in a layer above the current document).  If we can limit it to a single parent, I think it becomes simpler to manage.  Thoughts?